### PR TITLE
fix(core): skip git init when scaffolding inside an existing repo

### PR DESCRIPTION
**Pull request by lexicalunit**
## Why

`agent-native create <name>` run from inside an existing checkout leaves a git repo nested inside another one.

`finalizeScaffold` ends every scaffold at `tryGitInitUnlessRepo`, which checked only whether *the target directory itself* was already a repo:

```ts
function tryGitInitUnlessRepo(dir: string): void {
  if (fs.existsSync(path.join(dir, ".git"))) return;
  tryGitInit(dir);
}
```

A freshly created subdirectory never has its own `.git`, so the guard never fires and `tryGitInit` runs `git init` + `git add -A` + `git commit` inside it. The in-place path (`create .`) is already safe — its comment says the point is that "we never write an unexpected commit into the user's history" — but the named path had no equivalent guard.

A nested `.git` is a trap rather than a cosmetic issue. Finishing a generated workspace usually ends in `cp -a scaffold/. .`, which copies dotfiles and merges directories — so the scaffold's `.git/HEAD` lands on top of the parent's, silently moving the parent repo onto the scaffold's branch (`init.defaultBranch`, normally `main`) and taking the working tree with it. No git command the surrounding tooling can see ever runs.

That is not hypothetical: it cost two Builder Fusion projects a full session of generated code. Full write-up in [ENG-13640](https://builder-io.atlassian.net/browse/ENG-13640); the platform-side fix so a moved HEAD can no longer destroy work is [BuilderIO/ai-services#6061](https://github.com/BuilderIO/ai-services/pull/6061). This PR removes the trigger; that one removes the consequence. Both are worth having.

## What changed

- `tryGitInitUnlessRepo` now walks up from the target for the nearest enclosing `.git` instead of checking one directory. The walk starts at `dir`, so it **subsumes** the old same-directory check rather than stacking another special case beside it.
- Both scaffold paths already funnel through this one function, so the named and in-place cases are fixed together — no second guard.
- Behavior outside a checkout is unchanged: a scaffold with no repo above it still gets `git init` and the initial commit.
- The skip stays silent, matching what the in-place path has always done. `create.ts` has no console output of its own; the CLI wrapper prints.

A directory walk rather than `git rev-parse --is-inside-work-tree`: no subprocess, and no `catch` that would have to turn "git is missing" and "not a repo" into the same value, which `guard:no-silent-coercion` and the flagship rule in `AGENTS.md` both push against.

## Verification

- `pnpm vitest run packages/core/src/cli/create-e2e.spec.ts` — 77 passed, 1 skipped.
- `pnpm vitest run .../create-e2e.spec.ts .../create.spec.ts` — 115 passed.
- Two new cases next to the existing in-place git test: a scaffold inside a repo gets no nested `.git` and leaves the parent's HEAD, branch and status untouched; a scaffold outside any repo still gets its own repo and initial commit.
- Confirmed the first is a real regression guard — reverting the one-line change fails it with `expected true to be false`.
- `pnpm fmt:check` clean, `pnpm oxlint` reports nothing in the touched files.
- `pnpm guards`: everything passes except `guard:i18n-catalogs` and `guard:migration-manifest`, which fail identically on a stashed clean tree (composer symbol moves and catalog debt, unrelated to this change).
- `pnpm --filter @agent-native/core typecheck`: 123 pre-existing errors, byte-identical with and without this change, none in `cli/create`.

### CODE REVIEW

File: packages/core/src/cli/create.ts [hunk 1]
```diff
@@ -1094,10 +1093,29 @@ function finalizeScaffold(scaffoldDir: string, inPlace?: boolean): void {
...
+ * repo nested in another one. Copying such a scaffold into its parent — the
+ * `cp -a scaffold/. .` that finishing a generated workspace usually ends with —
+ * drags `.git/HEAD` along and silently moves the parent onto the scaffold's
+ * branch, taking the working tree with it.
+ */
+function findEnclosingRepo(dir: string): string | undefined {
+  let current = path.resolve(dir);
+  for (;;) {
+    if (fs.existsSync(path.join(current, ".git"))) return current;
+    const parent = path.dirname(current);
```
[the remark this answers is not in the document]
Comment (lexicalunit) [reply packages/core/src/cli/create.ts]: Good catch — fixed in 52645e5db.

You are right that walking to the path root is not what git does, and the failure direction matters: being *more* eager than git to find an enclosing repo means silently skipping `git init` for a workspace git would never associate with that repo, so the scaffold comes out with no repository at all. That is a worse outcome than the nesting this PR set out to prevent.

`findEnclosingRepo` now stops when the parent sits on a different device, and crosses only when `GIT_DISCOVERY_ACROSS_FILESYSTEM` says to. The env var is parsed the way `git_env_bool` does — unset, empty, `0`, `false`, `no`, `off` are all false — rather than a bare truthiness check, so `GIT_DISCOVERY_ACROSS_FILESYSTEM=0` behaves as git would treat it.

On the regression test: a real mount is not feasible in the unit suite, so the boundary is simulated by stubbing `fs.statSync` to report a different `dev` below a chosen directory. That covers the mechanism this fix turns on without needing privileges. Five cases now sit in `create.spec.ts` — repo above the target, target itself, no repo anywhere, stops at the boundary, and crosses it when the env var is set — and I confirmed the boundary case is a real guard by removing the device check and watching only that test fail.

`GIT_DISCOVERY_ACROSS_FILESYSTEM` is now documented in `docs/environment-variables.md`; `guard:env-documentation` caught that omission and passes. Full `guards` run is clean apart from `guard:i18n-catalogs`, which fails identically on a stashed clean tree.

File: packages/core/src/cli/create.ts [hunk 2]
```diff
@@ -1094,10 +1093,50 @@ function finalizeScaffold(scaffoldDir: string, inPlace?: boolean): void {
...
+ * The walk stops at a filesystem boundary because git's does. Climbing past one
+ * would claim a bind-mounted or otherwise separate workspace belongs to a repo
+ * git would never associate it with, leaving that scaffold with no repository
+ * of its own.
+ */
+function findEnclosingRepo(dir: string): string | undefined {
+  const crossesFilesystems = gitEnvBool(
+    process.env.GIT_DISCOVERY_ACROSS_FILESYSTEM,
+  );
+  let current = path.resolve(dir);
```
[the remark this answers is not in the document]
Comment (lexicalunit) [reply packages/core/src/cli/create.ts]: Fixed in 1ba86fb8f, by deleting the hand-rolled discovery rather than adding another rule to it.

All three findings this round, plus the filesystem-boundary one before them, were the same root cause: I was reimplementing git's repository discovery. Each fix would have been correct and the next gap would still have been waiting — and every one of them failed in the same direction, too eager to claim an enclosing repo, which is the direction that silently leaves a scaffold with no repository at all. Your third comment named the right answer: delegate to git.

`findEnclosingRepo` is now `git rev-parse --show-toplevel` run with `cwd: dir`. That resolves symlinked paths, honours filesystem boundaries, `GIT_CEILING_DIRECTORIES` and `GIT_DISCOVERY_ACROSS_FILESYSTEM`, parses those values exactly as git does, and stays right as git changes. `gitEnvBool` and the device walk are gone. Running git costs nothing here, since the alternative to finding a repo is shelling out to `git init` anyway.

Net −40 lines, and the tests exercise real git instead of a stubbed `fs.statSync`: a symlinked path into a repo subdirectory, and `GIT_CEILING_DIRECTORIES` set between the target and the repo. Both would have failed before this commit.

File: packages/core/src/cli/create.ts [hunk 3]
```diff
@@ -1094,10 +1093,50 @@ function finalizeScaffold(scaffoldDir: string, inPlace?: boolean): void {
...
 function tryGitInitUnlessRepo(dir: string): void {
-  if (fs.existsSync(path.join(dir, ".git"))) return;
+  if (findEnclosingRepo(dir)) return;
   tryGitInit(dir);
 }
 
+/** Mirrors git's `git_env_bool`: unset, empty, and the off words are false. */
+function gitEnvBool(value: string | undefined): boolean {
+  if (value === undefined) return false;
+  return !["", "0", "false", "no", "off"].includes(value.trim().toLowerCase());
```
[the remark this answers is not in the document]
Comment (lexicalunit) [reply packages/core/src/cli/create.ts]: Fixed in 1ba86fb8f, by deleting the hand-rolled discovery rather than adding another rule to it.

All three findings this round, plus the filesystem-boundary one before them, were the same root cause: I was reimplementing git's repository discovery. Each fix would have been correct and the next gap would still have been waiting — and every one of them failed in the same direction, too eager to claim an enclosing repo, which is the direction that silently leaves a scaffold with no repository at all. Your third comment named the right answer: delegate to git.

`findEnclosingRepo` is now `git rev-parse --show-toplevel` run with `cwd: dir`. That resolves symlinked paths, honours filesystem boundaries, `GIT_CEILING_DIRECTORIES` and `GIT_DISCOVERY_ACROSS_FILESYSTEM`, parses those values exactly as git does, and stays right as git changes. `gitEnvBool` and the device walk are gone. Running git costs nothing here, since the alternative to finding a repo is shelling out to `git init` anyway.

Net −40 lines, and the tests exercise real git instead of a stubbed `fs.statSync`: a symlinked path into a repo subdirectory, and `GIT_CEILING_DIRECTORIES` set between the target and the repo. Both would have failed before this commit.

File: packages/core/src/cli/create.ts [hunk 4]
```diff
@@ -1094,10 +1093,50 @@ function finalizeScaffold(scaffoldDir: string, inPlace?: boolean): void {
...
+ * `cp -a scaffold/. .` that finishing a generated workspace usually ends with —
+ * drags `.git/HEAD` along and silently moves the parent onto the scaffold's
+ * branch, taking the working tree with it.
+ *
+ * The walk stops at a filesystem boundary because git's does. Climbing past one
+ * would claim a bind-mounted or otherwise separate workspace belongs to a repo
+ * git would never associate it with, leaving that scaffold with no repository
+ * of its own.
+ */
+function findEnclosingRepo(dir: string): string | undefined {
```
[the remark this answers is not in the document]
Comment (lexicalunit) [reply packages/core/src/cli/create.ts]: Fixed in 1ba86fb8f, by deleting the hand-rolled discovery rather than adding another rule to it.

All three findings this round, plus the filesystem-boundary one before them, were the same root cause: I was reimplementing git's repository discovery. Each fix would have been correct and the next gap would still have been waiting — and every one of them failed in the same direction, too eager to claim an enclosing repo, which is the direction that silently leaves a scaffold with no repository at all. Your third comment named the right answer: delegate to git.

`findEnclosingRepo` is now `git rev-parse --show-toplevel` run with `cwd: dir`. That resolves symlinked paths, honours filesystem boundaries, `GIT_CEILING_DIRECTORIES` and `GIT_DISCOVERY_ACROSS_FILESYSTEM`, parses those values exactly as git does, and stays right as git changes. `gitEnvBool` and the device walk are gone. Running git costs nothing here, since the alternative to finding a repo is shelling out to `git init` anyway.

Net −40 lines, and the tests exercise real git instead of a stubbed `fs.statSync`: a symlinked path into a repo subdirectory, and `GIT_CEILING_DIRECTORIES` set between the target and the repo. Both would have failed before this commit.

**Review by kylefowler — APPROVED** [review]

File: packages/core/src/cli/create.ts [hunk 5]
```diff
@@ -1094,10 +1093,31 @@ function finalizeScaffold(scaffoldDir: string, inPlace?: boolean): void {
...
+ *
+ * Matching a `.git` entry by hand looks equivalent and is not: discovery also
+ * turns on symlinked paths, filesystem boundaries, `GIT_CEILING_DIRECTORIES`
+ * and `GIT_DISCOVERY_ACROSS_FILESYSTEM`, and getting any of them wrong in the
+ * over-eager direction silently leaves a scaffold with no repository at all.
+ * Running git is free here because the only alternative to finding a repo is
+ * shelling out to `git init` anyway.
+ */
+function findEnclosingRepo(dir: string): string | undefined {
+  const result = spawnSync("git", ["rev-parse", "--show-toplevel"], {
```
[the remark this answers is not in the document]
Comment (lexicalunit) [reply packages/core/src/cli/create.ts:1166]: Confirmed and fixed in 5599d493d. I reproduced it before changing anything — from a non-repo directory with `GIT_DIR` pointed elsewhere, `git rev-parse --show-toplevel` exits 0 and prints that directory as its work tree, so discovery would report a repo that is not there and the init would be skipped.

Discovery now runs with the repository-location overrides stripped: `GIT_DIR`, `GIT_WORK_TREE`, `GIT_INDEX_FILE`, `GIT_OBJECT_DIRECTORY`, `GIT_ALTERNATE_OBJECT_DIRECTORIES`, `GIT_COMMON_DIR`, `GIT_NAMESPACE`. `GIT_CEILING_DIRECTORIES` and `GIT_DISCOVERY_ACROSS_FILESYSTEM` are deliberately kept, as you noted — where the search stops is genuinely the caller's to configure.

I also fixed the sibling you did not flag: `tryGitInit` inherited the same overrides, and there the damage is worse. With `GIT_DIR` set, its `init`, `add -A` and initial commit all land in the ambient repository rather than the scaffold — so the failure mode was not just a missing repo but a stray commit in someone else's. Both now share one scrubbed env.

Two regression tests, both confirmed to fail on the previous commit: a unit test asserting discovery ignores an unrelated inherited `GIT_DIR`, and an e2e asserting a scaffold created under one still gets its own repository and initial commit while the unrelated repo stays at zero commits.

File: packages/core/src/cli/create.ts [hunk 6]
```diff
@@ -1094,10 +1093,57 @@ function finalizeScaffold(scaffoldDir: string, inPlace?: boolean): void {
...
+function envWithoutRepoOverrides(): NodeJS.ProcessEnv {
+  const env = { ...process.env };
+  for (const key of [
+    "GIT_DIR",
+    "GIT_WORK_TREE",
+    "GIT_INDEX_FILE",
+    "GIT_OBJECT_DIRECTORY",
+    "GIT_ALTERNATE_OBJECT_DIRECTORIES",
+    "GIT_COMMON_DIR",
+    "GIT_NAMESPACE",
```
[the remark this answers is not in the document]
Comment (lexicalunit) [reply packages/core/src/cli/create.ts]: Both fixed in e7500476b — and the fact that these were the third and fourth variables reported one at a time is the tell that my *list* was the problem, not any entry in it.

`envWithoutRepoOverrides` now mirrors git's own `local_repo_env`: the set git clears when it runs a command against a different repository, plus the `GIT_QUARANTINE_PATH` that `receive-pack` exports around hooks. Both variables you flagged are in it, along with `GIT_CONFIG*`, `GIT_GRAFT_FILE`, `GIT_IMPLICIT_WORK_TREE`, `GIT_NO_REPLACE_OBJECTS`, `GIT_PREFIX`, `GIT_REPLACE_REF_BASE` and `GIT_SHALLOW_FILE`, which had not come up yet. The rule is git's now rather than mine, which is the point. The discovery controls stay out of the set deliberately: where the search stops is the caller's to configure, unlike which repository it lands in.

Coverage moved up a level with it. Instead of one variable per test, the e2e scaffolds with `GIT_DIR`, `GIT_INDEX_FILE`, `GIT_OBJECT_DIRECTORY`, `GIT_COMMON_DIR`, `GIT_QUARANTINE_PATH` and `GIT_INTERNAL_SUPER_PREFIX` all set at once, and asserts the scaffold gets its own repository and initial commit while the inherited repo stays at zero commits. I verified it is a real guard by removing just the two you reported: it fails with `fatal: not a git repository`.

One side effect worth a look in review: naming the constant after git's concept made `guard:env-documentation` notice all eighteen names, so `docs/environment-variables.md` gains a `GIT_*` family row. That is the accurate description — this repo never sets any of them, it only clears them before shelling out.

[the remark this answers is not in the document]
Comment (lexicalunit) [reply packages/core/src/cli/create.ts]: Both fixed in e7500476b — and the fact that these were the third and fourth variables reported one at a time is the tell that my *list* was the problem, not any entry in it.

`envWithoutRepoOverrides` now mirrors git's own `local_repo_env`: the set git clears when it runs a command against a different repository, plus the `GIT_QUARANTINE_PATH` that `receive-pack` exports around hooks. Both variables you flagged are in it, along with `GIT_CONFIG*`, `GIT_GRAFT_FILE`, `GIT_IMPLICIT_WORK_TREE`, `GIT_NO_REPLACE_OBJECTS`, `GIT_PREFIX`, `GIT_REPLACE_REF_BASE` and `GIT_SHALLOW_FILE`, which had not come up yet. The rule is git's now rather than mine, which is the point. The discovery controls stay out of the set deliberately: where the search stops is the caller's to configure, unlike which repository it lands in.

Coverage moved up a level with it. Instead of one variable per test, the e2e scaffolds with `GIT_DIR`, `GIT_INDEX_FILE`, `GIT_OBJECT_DIRECTORY`, `GIT_COMMON_DIR`, `GIT_QUARANTINE_PATH` and `GIT_INTERNAL_SUPER_PREFIX` all set at once, and asserts the scaffold gets its own repository and initial commit while the inherited repo stays at zero commits. I verified it is a real guard by removing just the two you reported: it fails with `fatal: not a git repository`.

One side effect worth a look in review: naming the constant after git's concept made `guard:env-documentation` notice all eighteen names, so `docs/environment-variables.md` gains a `GIT_*` family row. That is the accurate description — this repo never sets any of them, it only clears them before shelling out.

File: packages/core/src/cli/create.ts [hunk 7]
```diff
@@ -1094,10 +1093,75 @@ function finalizeScaffold(scaffoldDir: string, inPlace?: boolean): void {
...
+ * quarantine — may have pinned git to its own repository. Inheriting that makes
+ * discovery answer about that repository rather than the directory being
+ * scaffolded, and makes an init write into it. Picking the variables off one at
+ * a time invites the next gap, so mirror the list git clears for exactly this
+ * reason. The discovery controls are deliberately not in it: where the search
+ * stops really is the caller's to configure.
+ */
+function envWithoutRepoOverrides(): NodeJS.ProcessEnv {
+  const env = { ...process.env };
+  for (const key of GIT_LOCAL_REPO_ENV) delete env[key];
```
[the remark this answers is not in the document]
Comment (lexicalunit) [reply packages/core/src/cli/create.ts]: Both fixed in adbbad824. Taking them in turn, because they are opposite kinds of mistake.

**The config regression is mine, and it is exactly the over-correction it looks like.** Last round I generalised a hand-picked list into git's whole `local_repo_env`, which swept in `GIT_CONFIG`, `GIT_CONFIG_COUNT` and `GIT_CONFIG_PARAMETERS`. Those are not repository bindings: git clears them because it is entering a *different* repository and the caller's `-c` overrides were scoped to the old one, whereas here the caller is configuring the scaffold that is about to exist. `init.defaultBranch`, `core.hooksPath` and `commit.gpgsign` now survive. The set is named `GIT_REPO_LOCATION_ENV` and its comment states the narrowing, so the next reader does not "fix" it back.

**The discovery one is the more serious of the two**, and you are right that it turns this guard into its own opposite. `rev-parse` exits 128 for a checkout it refuses just as readily as for a repo-less directory, so collapsing every non-zero into "no repository" meant a dubious-ownership repo — routine in containers and CI — would get a nested `.git` created inside it. That is the precise failure this PR exists to prevent.

Discovery now returns `inside` / `outside` / `unknown`, and only `outside` initialises. `LC_ALL` is pinned for the child so the message the states are told apart by is not translated.

For deterministic coverage without needing a second uid, I used a corrupt gitfile: it produces `fatal: invalid gitfile format:`, which is a refusal that does not say "not a git repository" — the same shape as dubious ownership. Two tests, both confirmed to fail if the collapse is restored: a unit test asserting `unknown` rather than `outside`, and an e2e asserting a scaffold under such a parent gets no `.git`.

File: packages/core/src/cli/create.ts [hunk 8]
```diff
@@ -1094,10 +1093,75 @@ function finalizeScaffold(scaffoldDir: string, inPlace?: boolean): void {
...
+ */
+function findEnclosingRepo(dir: string): string | undefined {
+  const result = spawnSync("git", ["rev-parse", "--show-toplevel"], {
+    cwd: dir,
+    encoding: "utf8",
+    env: envWithoutRepoOverrides(),
+  });
+  // A non-zero exit is git answering "no work tree here". A spawn error means
+  // git is unusable, and an init attempt has nothing to add in that case.
+  if (result.error || result.status !== 0) return undefined;
```
[the remark this answers is not in the document]
Comment (lexicalunit) [reply packages/core/src/cli/create.ts]: Both fixed in adbbad824. Taking them in turn, because they are opposite kinds of mistake.

**The config regression is mine, and it is exactly the over-correction it looks like.** Last round I generalised a hand-picked list into git's whole `local_repo_env`, which swept in `GIT_CONFIG`, `GIT_CONFIG_COUNT` and `GIT_CONFIG_PARAMETERS`. Those are not repository bindings: git clears them because it is entering a *different* repository and the caller's `-c` overrides were scoped to the old one, whereas here the caller is configuring the scaffold that is about to exist. `init.defaultBranch`, `core.hooksPath` and `commit.gpgsign` now survive. The set is named `GIT_REPO_LOCATION_ENV` and its comment states the narrowing, so the next reader does not "fix" it back.

**The discovery one is the more serious of the two**, and you are right that it turns this guard into its own opposite. `rev-parse` exits 128 for a checkout it refuses just as readily as for a repo-less directory, so collapsing every non-zero into "no repository" meant a dubious-ownership repo — routine in containers and CI — would get a nested `.git` created inside it. That is the precise failure this PR exists to prevent.

Discovery now returns `inside` / `outside` / `unknown`, and only `outside` initialises. `LC_ALL` is pinned for the child so the message the states are told apart by is not translated.

For deterministic coverage without needing a second uid, I used a corrupt gitfile: it produces `fatal: invalid gitfile format:`, which is a refusal that does not say "not a git repository" — the same shape as dubious ownership. Two tests, both confirmed to fail if the collapse is restored: a unit test asserting `unknown` rather than `outside`, and an e2e asserting a scaffold under such a parent gets no `.git`.
